Utilities/GCDMulticastDelegate.h

```diff
@@ -55,4 +55,43 @@ NS_ASSUME_NONNULL_BEGIN
 - (BOOL)getNextDelegate:(id _Nullable * _Nonnull)delPtr delegateQueue:(dispatch_queue_t _Nullable * _Nonnull)dqPtr forSelector:(SEL)aSelector;
 
 @end
+
+/**
+ * A helper class for propagating custom context across multicast delegate invocations.
+ *
+ * This class serves 2 main purposes:
+ * - provides an auxiliary path of custom data delivery to the invoked delegate methods
+ * - makes it possible to track the delegate method invocation completion
+ *
+ * The context propagates along the cascade of invocations, i.e. when a delegate method calls another multicast delegate,
+ * that subsequent invocation belongs to the same context. This is particularly relevant w.r.t. the xmpp framework
+ * architecture where a scenario with two layers of delegation is common: stream -> module and module -> application.
+ * The propagating context is what enables the framework to deliver stream event-related information across modules
+ * to the application callbacks.
+ *
+ * The default context propagation junctions are invocation forwarding and delegate enumerator creation. As long as
+ * they are executed under an existing context, the propagation is automatic.
+ *
+ * A manual propagation scenario (e.g. asynchronous message processing within a module) would consist of the following steps:
+ * 1. Capturing the context object while still in the delegate callback context with @c currentContext.
+ * 2. Entering the captured context's @c continuityGroup.
+ * 3. Restoring the context on an arbitrary queue with @c becomeCurrentOnQueue:forActionWithBlock:
+ * 4. Leaving the @c continuityGroup within the action block submitted to @c becomeCurrentOnQueue:forActionWithBlock:
+ *
+ * Steps 2. and 4. are only required if @c becomeCurrentOnQueue:forActionWithBlock: itself is invoked asynchronously
+ * (e.g. in a network or disk IO completion callback).
+ */
+@interface GCDMulticastDelegateInvocationContext : NSObject
```

Comment:
I can see why this is needed for the current approach but it feels complex and looks like it requires enforcing a certain style to future XMPPStream changes. For instance now instead of calling `multicastDelegate` directly, you have to be sure to call  `performDelegateActionWithElementEvent`, and if you don't... what happens?

Reply:
See [my comment](https://github.com/robbiehanson/XMPPFramework/pull/996?_pjax=%23js-repo-pjax-container#issuecomment-345440177) below.